dali/pipeline/operator/arg_helper.h

```diff
@@ -15,96 +15,158 @@
-  inline bool IsSet() const noexcept { return is_set_; }
-
-  template <int ndim = DynamicDimensions>
-  inline const TensorView<StorageCPU, T, ndim> Tensor(int sample_index) const {
-    ASSERT(IsInput());
-    return view<T, ndim>(arg_input_[sample_index]);
+  bool IsConstant() const {
+    return has_arg_const_;
   }
 
-  inline const T &Value() const {
-    return value_;
+  bool IsArgInput() const {
+    return has_arg_input_;
   }
 
-  inline const string &Name() const noexcept {
-    return name_;
+  void Acquire(const OpSpec &spec, const ArgumentWorkspace &ws, int nsamples,
+               const TensorShape<ndim> &expected_shape) {
+    if (has_arg_input_) {
+      view_ = view<const T, ndim>(ws.ArgumentInput(arg_name_));
+      DALI_ENFORCE(is_uniform(view_.shape) && expected_shape == view_.shape[0],
+        make_string("Expected uniform shape for argument \"", arg_name_,
+                    "\" but got shape ", view_.shape));
+    } else if (has_arg_const_) {
+      TensorShape<ndim> shape{};
+      if (ndim == 0) {
+        data_ = {spec.GetArgument<T>(arg_name_)};
+      } else {
+        assert (ndim <= 1);
+        data_ = spec.GetRepeatedArgument<T>(arg_name_);
+        int64_t len = data_.size();
+        int64_t expected_len = volume(expected_shape);
+        if (len == 1 && expected_shape > 1) {
+          data_.resize(expected_len, data_[0]);
+        } else {
+          DALI_ENFORCE(len == volume(expected_shape),
+            make_string("Argument \"", arg_name_, "\" expected shape ", expected_shape,
+                        " but got ", len, " values, which can't be interpreted as the expected shape."));
+        }
+      }
+      view_ = constant_view(nsamples, data_.data(), shape);
+    }
   }
 
-  inline const T &operator[](int sample_index) const {
-    if (IsInput()) {
-#if DALI_DEBUG
-      DALI_ENFORCE(sample_index < static_cast<int>(arg_input_->ntensor()));
-#endif
-      return (*arg_input_)[sample_index].data<T>()[0];
+  void Acquire(const OpSpec &spec, const ArgumentWorkspace &ws, int nsamples,
+               bool enforce_uniform = false) {
+    if (has_arg_input_) {
+      view_ = view<const T, ndim>(ws.ArgumentInput(arg_name_));
+      if (enforce_uniform) {
+        DALI_ENFORCE(is_uniform(view_.shape),
+          make_string("Expected uniform shape for argument \"", arg_name_,
+                      "\" but got shape ", view_.shape));
+      }
     } else {
-      return  value_;
+      TensorShape<ndim> shape{};
+      if (ndim == 0) {
+        data_ = {spec.GetArgument<T>(arg_name_)};
+      } else if (ndim == 1) {
+        data_ = spec.GetRepeatedArgument<T>(arg_name_);
+        int64_t len = data_.size();
+        shape = std::array<int64_t, 1>{len};
+      } else {
+        // ndim > 1 but we don't have information about the expected shape.
+        // An overload with ``expected_shape`` should have been used.
+        assert(false);
```

Comment:
I guess sometimes we could guess the shape from size. We could pass a function:
```C++
template <typename ndim>
struct ArgShapeFromSize {
  TensorShape<ndim> operator(int64_t size) const {
    throw std::logic_error("Cannot infer shape from flat size.");
  }
};

template <>
struct ArgShapeFromSize<0> {
  TensorShape<ndim> operator(int64_t size) const {
    DALI_ENFORCE(size == 1, make_string("Expected a scalar argument, got ", size, " values));
    return {};
  }
};

template <>
struct ArgShapeFromSize<1> {
  TensorShape<1> operator(int64_t size) const {
    return { size };
  }
};


  template <typename ShapeFromSizeFn = ArgShapeFromSize<ndim> >
  void Acquire(const OpSpec &spec, const ArgumentWorkspace &ws, int nsamples, bool enforce_uniform = false,
  ShapeFromSizeFn &&shape_from_size = {}) {
```

With that, we could have, for example, something for shear matrix or affine transforms:
```C++
matrix.Acquire(spec_, ws, batch_size, true, [&](int64_t size) {
  int n = sqrt(size);
  DALI_ENFORCE(size == n*(n+1), make_string("Cannot form an affine transform matrix with ", size, " elements"));
  return { n, n+1 };
});
```

Reply:
Good idea, thanks.